src/sdp/parser.rs

```diff
@@ -232,55 +239,71 @@ where
+        }),
+        not_sp(),
+        token(' '),
+        not_sp().and_then(|s| {
+            s.parse::<u16>()
+                .map_err(StreamErrorFor::<Input>::message_format)
+        }),
+        token(' '),
+        not_sp().and_then(|s| {
+            s.as_str().try_into().map_err(|_| {
+                StreamErrorFor::<Input>::message_format(format!("invalid protocol: {}", s))
+            })
+        }),
+        token(' '),
+        not_sp().and_then(|s| {
+            s.parse::<u32>()
+                .map_err(StreamErrorFor::<Input>::message_format)
+        }),
+        token(' '),
+        ip_addr(),
+        token(' '),
+        port(),
+        string(" typ "),
+        kind,
+        optional((
+            attempt(string(" raddr ")),
             ip_addr(),
-            token(' '),
+            string(" rport "),
             port(),
-            string(" typ "),
-            kind,
-            optional((
-                attempt(string(" raddr ")),
-                ip_addr(),
-                string(" rport "),
-                port(),
-            )),
-            optional((attempt(string(" ufrag ")), not_sp())),
-        ),
-    )
-    .map(
-        |(found, _, comp_id, _, proto, _, prio, _, addr, _, port, _, kind, raddr, ufrag)| {
-            Candidate::parsed(
-                found,
-                comp_id,
-                proto,
-                prio, // remote candidates calculate prio on their side
-                SocketAddr::from((addr, port)),
-                kind,
-                raddr.map(|(_, addr, _, port)| SocketAddr::from((addr, port))),
-                ufrag.map(|(_, u)| u),
-            )
-        },
+        )),
+        optional((attempt(string(" ufrag ")), not_sp())),
     )
+        .map(
+            |(_, found, _, comp_id, _, proto, _, prio, _, addr, _, port, _, kind, raddr, ufrag)| {
+                Candidate::parsed(
+                    found,
+                    comp_id,
+                    proto,
+                    prio, // remote candidates calculate prio on their side
+                    SocketAddr::from((addr, port)),
+                    kind,
+                    raddr.map(|(_, addr, _, port)| SocketAddr::from((addr, port))),
+                    ufrag.map(|(_, u)| u),
+                )
+            },
+        )
+}
+
+/// Parser for a=candidate lines.
+#[doc(hidden)]
+pub fn candidate<Input>() -> impl Parser<Input, Output = Candidate>
+where
+    Input: Stream<Token = char>,
+    Input::Error: ParseError<Input::Token, Input::Range, Input::Position>,
+{
+    // a=candidate:1 1 udp 2113929471 203.0.113.100 10100 typ host
+    // a=candidate:1 2 udp 2113929470 203.0.113.100 10101 typ host
+    // a=candidate:1 1 udp 1845494015 198.51.100.100 11100 typ srflx raddr 203.0.113.100 rport 10100
+    // a=candidate:1 1 udp 255 192.0.2.100 12100 typ relay raddr 198.51.100.100 rport 11100
+
+    (string("a="), candidate_attribute()).map(|(_, c)| c)
```

Comment:
I believe you can still reuse the `attribute_line` parser function outermost here.

Reply:
Unfortunately `attribute_line` still needs the name of the attribute which is already handled.